### PR TITLE
[luci-interpreter] Implement check_input_tensor_dimensions

### PR DESCRIPTION
**Pull request by seanshpark**
This will fill implementation of check_input_tensor_dimensions() method.

ONE-DCO-1.0-Signed-off-by: SaeHie Park <saehie.park@gmail.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
for #9895 from #9899
- code itself is from TensorFlow lite kernel

**Comment by seanshpark** [comment]
cc @mhs4670go

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 1]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+  //  > 0 means clipping
+  LUCI_INTERPRETER_CHECK(params().cell_clip >= 0);
+  LUCI_INTERPRETER_CHECK(params().proj_clip >= 0);
+
+  if (input_to_input_weights() != nullptr)
+  {
+    const Shape &input_to_input_weights_shape = input_to_input_weights()->shape();
+    LUCI_INTERPRETER_CHECK(input_to_input_weights_shape.num_dims() == 2);
+    LUCI_INTERPRETER_CHECK(input_to_input_weights_shape.dim(0) == n_cell);
+    LUCI_INTERPRETER_CHECK(input_to_input_weights_shape.dim(1) == n_input);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:102]: Original TF code has type check

```
    TF_LITE_ENSURE_TYPES_EQ(context, input_to_input_weights->type,
                            input_to_forget_weights->type);
```

Is it ok to skip the check?

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 2]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+  }
+
+  const Shape &input_to_forget_weights_shape = input_to_forget_weights()->shape();
+  LUCI_INTERPRETER_CHECK(input_to_forget_weights_shape.num_dims() == 2);
+  LUCI_INTERPRETER_CHECK(input_to_forget_weights_shape.dim(0) == n_cell);
+  LUCI_INTERPRETER_CHECK(input_to_forget_weights_shape.dim(1) == n_input);
+
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:109]: TF_LITE_ENSURE(context, (input_to_forget_weights->type == kTfLiteFloat32) ||
                              (input_to_forget_weights->type == kTfLiteUInt8) ||
                              (input_to_forget_weights->type == kTfLiteInt8));

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 3]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+  const Shape &input_to_cell_weights_shape = input_to_cell_weights()->shape();
+  LUCI_INTERPRETER_CHECK(input_to_cell_weights_shape.num_dims() == 2);
+  LUCI_INTERPRETER_CHECK(input_to_cell_weights_shape.dim(0) == n_cell);
+  LUCI_INTERPRETER_CHECK(input_to_cell_weights_shape.dim(1) == n_input);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:113]: TF_LITE_ENSURE_TYPES_EQ(context, input_to_cell_weights->type,
                          input_to_forget_weights->type);

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 4]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+
+  if (recurrent_to_input_weights() != nullptr)
+  {
+    const Shape &recurrent_to_input_weights_shape = recurrent_to_input_weights()->shape();
+    LUCI_INTERPRETER_CHECK(recurrent_to_input_weights_shape.num_dims() == 2);
+    LUCI_INTERPRETER_CHECK(recurrent_to_input_weights_shape.dim(0) == n_cell);
+    LUCI_INTERPRETER_CHECK(recurrent_to_input_weights_shape.dim(1) == n_output);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:120]: TF_LITE_ENSURE_TYPES_EQ(context, recurrent_to_input_weights->type,
                            input_to_forget_weights->type);

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 5]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+  }
+
+  const Shape &recurrent_to_forget_weights_shape = recurrent_to_forget_weights()->shape();
+  LUCI_INTERPRETER_CHECK(recurrent_to_forget_weights_shape.num_dims() == 2);
+  LUCI_INTERPRETER_CHECK(recurrent_to_forget_weights_shape.dim(0) == n_cell);
+  LUCI_INTERPRETER_CHECK(recurrent_to_forget_weights_shape.dim(1) == n_output);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:126]: TF_LITE_ENSURE_TYPES_EQ(context, recurrent_to_forget_weights->type,
                          input_to_forget_weights->type);

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 6]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+
+  const Shape &recurrent_to_cell_weights_shape = recurrent_to_cell_weights()->shape();
+  LUCI_INTERPRETER_CHECK(recurrent_to_cell_weights_shape.num_dims() == 2);
+  LUCI_INTERPRETER_CHECK(recurrent_to_cell_weights_shape.dim(0) == n_cell);
+  LUCI_INTERPRETER_CHECK(recurrent_to_cell_weights_shape.dim(1) == n_output);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:131]: TF_LITE_ENSURE_TYPES_EQ(context, recurrent_to_cell_weights->type,
                          input_to_forget_weights->type);

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp [hunk 7]
```diff
@@ -88,12 +88,251 @@ void UnidirectionalSequenceLSTM::check_input_tensor_dimensions(int n_input, int
...
+  {
+    LUCI_INTERPRETER_CHECK(output_gate_bias()->element_type() == loco::DataType::FLOAT32);
+  }
+
+  if (projection_weights() != nullptr)
+  {
+    const Shape &projection_weights_shape = projection_weights()->shape();
+    LUCI_INTERPRETER_CHECK(projection_weights_shape.num_dims() == 2);
+    LUCI_INTERPRETER_CHECK(projection_weights_shape.dim(0) == n_output);
+    LUCI_INTERPRETER_CHECK(projection_weights_shape.dim(1) == n_cell);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:243]: TF_LITE_ENSURE_TYPES_EQ(context, projection_weights->type,
                            input_to_forget_weights->type);

File: compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp (hunk 1 above)
Comment (seanshpark) [reply compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:102]: I've referenced TF2.8.0. which version did you refer?

Comment (jinevening) [reply compiler/luci-interpreter/src/kernels/UnidirectionalSequenceLSTM.cpp:102]: I'm seeing v2.10.0.

GitLens shows that the type checks were added about 2 years ago while the rest of the codes are written in 5 years ago.

**Comment by seanshpark** [comment]
@jinevening , as you've referenced 2.10 and I've done with 2.8, there will be differences.
Actually current luci-interpreter codes go with TF 2.6.0 kernels.
I'd like to go with TF2.8.0 as I had plan to upgrade to use with tflite 2.8.0 source (but fails with some test issue with ruy)

**Review by jinevening — APPROVED** [review]
LGTM

**Pull request merged**